rust/azure_iot_operations_services/src/azure_device_registry/client.rs

```diff
@@ -1,2 +1,1730 @@
+        let response = self
+            .notify_on_device_update_command_invoker
+            .invoke(unobserve_request)
+            .await
+            .map_err(ErrorKind::from)?;
+        match response.payload.notification_preference_response {
+            adr_name_gen::NotificationPreferenceResponse::Accepted => {
+                let receiver_id = Self::hash_device_endpoint(&device_name, &inbound_endpoint_name);
+                // Remove it from our dispatcher
+                if self
+                    .device_update_notification_dispatcher
+                    .unregister_receiver(&receiver_id)
+                {
+                    log::debug!(
+                        "Device `{device_name:?}` with inbound endpoint `{inbound_endpoint_name:?}` removed from observed list"
+                    );
+                } else {
+                    log::debug!(
+                        "Device `{device_name:?}` with inbound endpoint `{inbound_endpoint_name:?}` not in observed list"
+                    );
+                }
+                Ok(())
+            }
+            adr_name_gen::NotificationPreferenceResponse::Failed => {
+                Err(Error(ErrorKind::ObservationError))
+            }
+        }
+    }
+
+    /// Hashes the device name and inbound endpoint name to create a single string.
+    fn hash_device_endpoint(device_name: &str, inbound_endpoint_name: &str) -> String {
+        // `~`` can't be in a topic token, so this will never collide with another device + inbound endpoint name combo
+        format!("{device_name}~{inbound_endpoint_name}")
+    }
+
+    /// Unhashes the device name and inbound endpoint name from a single string.
+    fn un_hash_device_endpoint(hashed_device_endpoint: &str) -> Option<(String, String)> {
+        hashed_device_endpoint
+            .split_once('~')
+            .map(|(device_name, inbound_endpoint_name)| {
+                (device_name.to_string(), inbound_endpoint_name.to_string())
+            })
+    }
+
+    // ~~~~~~~~~~~~~~~~~ Asset APIs ~~~~~~~~~~~~~~~~~~~~~
+
+    /// Retrieves an asset from a Azure Device Registry service.
+    ///
+    /// # Arguments
+    /// * `device_name` - The name of the Device.
+    /// * `inbound_endpoint_name` - The name of the inbound endpoint.
+    /// * `asset_name` - The name of the asset.
+    /// * `timeout` - The duration until the Client stops waiting for a response to the request, it is rounded up to the nearest second.
+    ///
+    /// Returns a [`Asset`] if the the asset was found.
+    ///
+    /// # Errors
+    /// [`struct@Error`] of kind [`InvalidRequestArgument`](ErrorKind::InvalidRequestArgument)
+    /// if there is error building the request.
+    ///
+    /// [`struct@Error`] of kind [`AIOProtocolError`](ErrorKind::AIOProtocolError)
+    /// if topic tokens are not found.
+    ///
+    /// [`struct@Error`] of kind [`AIOProtocolError`](ErrorKind::AIOProtocolError)
+    /// if topic tokens are not found.
```

Comment:
```suggestion
    /// [`struct@Error`] of kind [`AIOProtocolError`](ErrorKind::AIOProtocolError)
    /// if device, inbound_endpoint, or asset names are invalid.
```
duplicate. Also, topic tokens don't mean anything to the caller of this fn, so be more descriptive about what can cause the error. From these descriptions, I'm not sure what causes this error vs InvalidRequestArgument either

Reply:
please apply this comment to the docs for all of the other functions - the error info needs to be actionable for the caller